File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    
    static String[] answer;
    static ArrayList<String> list;
    static boolean found;
    
    public String[] solution(String[][] tickets) {
        // 그래프 생성 및 두 번째 값을 기준으로 정렬
        Arrays.sort(tickets, new Comparator<String[]>() {
            @Override
            public int compare(String[] a, String[] b) {
                return a[1].compareTo(b[1]);
            }
        });
        
        boolean[] visited = new boolean[tickets.length];
        list = new ArrayList<>();
        list.add("ICN");
        found = false;
        
        dfs(tickets, visited, "ICN", 0);
        
        answer = new String[list.size()];
        answer = list.toArray(answer);
        return answer;
    }
    private static void dfs(String[][] tickets, boolean[] visited, String str, int count){
        //모든 티켓을 다 사용한 경우
        if(count == tickets.length) {
            found = true;
            return;
        }
        
        for(int i=0; i<tickets.length; i++){
            if(str.equals(tickets[i][0]) && !visited[i]){
                list.add(tickets[i][1]);
                visited[i] = true;
                dfs(tickets, visited, tickets[i][1], count+1);
                if(found) return;
                list.remove(list.size()-1);
                visited[i] = false;
            }
        }
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    tickets = sorted(tickets, key=lambda x:(x[0], x[1]))
    N = len(tickets)
    path = ["ICN"]
    visited = [False] * N
    
    def dfs(current):
        if len(path) == N+1:
            answer.append(path[:])
            return True
        
        for i in range(N):
            if not visited[i] and tickets[i][0] == current:
                visited[i] = True
                path.append(tickets[i][1])
                if dfs(tickets[i][1]):
                    return
                visited[i] = False
                path.pop()
    
    
    dfs("ICN")

    return answer[0]